rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,243 @@
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+```
+
+An alternative solution would be to extend mechanism `attr::scales`. This
+solution doesn't require new API, but extends the existing one.
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+```
+
+While the alternative solution allows to avoid new API, it introduces one more
+way to set scale which complicates both implementation and integration.
+Please review section
+[Unifying scaling in long-term](#3-unifying-scaling-in-long-term) for more
+details.
+
+
+### 2.3. Scales for inputs in Concat
+
+The mechanism `attr:set_scales` will be introduced to Concat primitive. This
+mechanism is used in Binary primitive, so API is already there.
+
+The following limitations are in place for an initial implementation:
+ - no-runtime support;
+ - mask is 0.
+
+An example:
+
+```cpp
+    primitive_attr attr;
+    for (int n = 0; n < num_src; ++n)
+        attr.set_scales(DNNL_ARG_MULTIPLE_SRC + n, /* mask */ 0, src_scales[n]);
+```
+
+## 3. Unifying scaling in long-term
+
+The motivation behind unifying scaling:
+- Remove duplications;
+- Provide an aligned support across all use cases.
+
+There are a few issues with the current scaling support:
+ - Output scale, scales and binary_mul overlap each other;
+ - RNN quantization is specific to RNN, but can be easily replaced by other
+ scaling mechanisms;
+ - Built-in parameter provides very limited flexibility (no mask, no runtime).
+
+The changes:
+1. A unified scaling will be supported using mechanism `attr::set_scales` only:
+ - `attr::output_scales`
+ ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+ ```
+ - `attr::p_op::binary_mul`
+ ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+ ```
+ - `attr::rnn_data_qparams`
+ ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_SRC, mask, data_scales);
+    attr.set_scales(DNNL_ARG_DST, mask, data_scales);
+ ```
+ - `attr::rnn_weights_qparams`
+ ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_WEIGHTS, mask, weights_scales);
+ ```
+ - `attr::p_op::built-in parameter`
+ ``` cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+    p_ops.append_eltwise(algorithm::eltwise_relu, alpha, beta);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, /* binary scale */ {DNNL_RUNTIME_F32_VAL});
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_DST,
+            /* mask */ 0, /* eltwise scale */ {DNNL_RUNTIME_F32_VAL});
+ ```
+2. Creation-time scaling support will be dropped.
```

Comment:
Makes sense. I will add new API without a mask.

Current masks are abstracted from primitive type similar to format tags. For example, reduction can be executed on memory with tag `abc` and the current mask can be used to specify scaling. in oneDNN there are a few primitives that work with `abcd` and don't have notion of batch, channels, etc. However we could pre-define some set of masks frequently used by users:
```cpp
#define DNNL_MASK_PER_DIM_0 1 << 0
#define DNNL_MASK_PER_DIM_1 1 << 1
...
```

Reply:
... It would also not make @nivas-x86 and I scratch our heads for a while trying to remember what `mask = 2` means for reorders w/quantization (we also don't have it well documented apparently).

Reply:
My point was that user mostly rely on per channel and per tensor. For the per channel,  they have to apply the following logic:
- for all conv tensors, C is the dimensions 1, so mask is 1 << 1.
- for rnn or layernorm, we have tnc tensors, for which c is the dimension 2, so the mask is 1 << 2.

Hence, I think it would simplify usage if user specify their intent independently of descriptor (like flags), instead of a bit mask where each bit meaning depends on descriptor.

Reply:
@mgouicem @msotoflo Please review a new section in the document: https://github.com/igorsafo/oneDNN/tree/rfcs/rfcs/20220201-quantization-scaling#simplifying-scales-configuration